### PR TITLE
To refactor Sparse Adam algorithm for functional form

### PR DESCRIPTION
**Pull request by iramazanli**
Adds Functional Interface for Sparse Adam Optimizer.

### CODE REVIEW

File: torch/optim/sparse_adam.py [hunk 1]
```diff
@@ -64,59 +64,48 @@ def step(self, closure=None):
                 loss = closure()
 
         for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            exp_avgs = []
+            exp_avg_sqs = []
+            state_steps = []
```
Comment (vincentqb) [inline torch/optim/sparse_adam.py:71]: that's a lot of memory that is cumulated: one item per element of group['params']. is there any other way?

Comment (vincentqb) [reply torch/optim/sparse_adam.py:71]: as you pointed out, adam also does this since #44791

Comment (vincentqb) [reply torch/optim/sparse_adam.py:71]: as discussed with @iramazanli, `state = self.state[p]` means the memory is saved anyway.

**Review by vincentqb — APPROVED** [review]
LGTM, thanks for the clarification.

don't forget to update the description :)

File: torch/optim/sparse_adam.py (hunk 1 above)
Comment (iramazanli) [reply torch/optim/sparse_adam.py:71]: it was a nice discussion :) thanks for pointing this and discussing with me.

**Pull request closed without merging**